lib/routes/token.js

```diff
@@ -89,6 +229,75 @@ function confirmRefreshToken(params) {
   });
 }
 
+function confirmJwt(params) {
+  var assertion = params.assertion;
+  logger.debug('jwt.confirm', assertion);
+
+  return JWTOOL.verify(assertion).catch(function(err) {
+    logger.info('jwt.invalid.verify', err.message);
+    throw AppError.invalidAssertion();
+  }).then(function(payload) {
+    logger.verbose('jwt.payload', payload);
+
+    // this cannot fail! huh, why?
+    // if the assertion couldn't decode, or the jku was not in our
+    // trusted list, the assertion would not have verified.
+    var client = SERVICE_CLIENTS[JwTool.unverify(assertion).header.jku];
+
+    // ohai eslint complexity
+    var uid = _validateJwtSub(payload.sub);
+
+    if (payload.aud !== JWT_AUD) {
+      logger.debug('jwt.invalid.aud', payload.aud);
+      throw AppError.invalidAssertion();
+    }
+
+    if (!Scope(client.scope).has(payload.scope)) {
+      logger.debug('jwt.invalid.scopes', {
+        allowed: client.scope,
+        requested: payload.scope
+      });
+      throw AppError.invalidScopes(payload.scope);
+    }
+
+    var now = Date.now() / 1000;
+    if ((payload.iat || Infinity) > now) {
+      logger.debug('jwt.invalid.iat', { now: now, iat: payload.iat });
+      throw AppError.invalidAssertion();
+    }
+    if ((payload.exp || -Infinity) < now) {
+      logger.debug('jwt.invalid.exp', { now: now, exp: payload.exp });
+      throw AppError.invalidAssertion();
+    }
+
+    return {
+      clientId: client.id,
+      userId: uid,
+      scope: payload.scope,
+    };
+  });
+}
+
+function _validateJwtSub(sub) {
+  var parts = (sub || '').split('@');
+  var uid = parts[0];
+  var ns = parts[1];
+  var allowedNs = config.get('browserid.issuer');
+  if (!uid) {
+    logger.debug('jwt.invalid.sub.missing');
+    throw AppError.invalidAssertion();
+  }
+  if (uid.length !== 32 ||
+    !HEX_STRING.test(uid) ||
+    ns !== allowedNs ||
+    parts.length !== 2) {
+    logger.debug('jwt.invalid.sub', sub);
+    throw AppError.invalidAssertion();
+  }
+
+  return uid;
+}
```

Comment:
After sleeping on it, I lean towards allowing raw FxA uids here rather than requiring them to be namespaced.  It's the form we've been encouraging clients and reliers to use in other places.